test/library.cpp

```diff
@@ -327,28 +411,294 @@ class LibraryTest : public ::testing::Test {
+    EXPECT_EQ(onlyValidBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[5].getBookId(), bookId1+"_date_flavour");
+
+
+    ASSERT_EQ(allBookmarks.size(), 6);
+    EXPECT_EQ(allBookmarks[0].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[1].getBookId(), bookId1);
+    EXPECT_EQ(allBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[5].getBookId(), bookId1+"_date_flavour");
+}
+
+TEST_F(LibraryTest, GetBestTargetBookIdOlder)
+{
+    auto bookId = std::string("0c45160e-f917-760a-9159-dfe3c53cdcdd");
+
+    auto book = lib->getBookById(bookId);
+
+    auto validBookmark = createBookmark(book);
+    lib->addBookmark(validBookmark);
+
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::UPGRADE_ONLY), bookId+"_date");
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::ALLOW_DOWNGRADE), bookId+"_date");
+}
+
+TEST_F(LibraryTest, GetBestTargetBookIdNewer)
+{
+    auto bookId = std::string("0c45160e-f917-760a-9159-dfe3c53cdcdd_date");
+
+    auto book = lib->getBookById(bookId);
+    EXPECT_EQ(book.getDate(), "2019-10-08");
+
+    auto validBookmark = createBookmark(book);
+    validBookmark.setDate("2020-10-08");
+    lib->addBookmark(validBookmark);
+
+    //  The best book for the bookmark is bookId...
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::UPGRADE_ONLY), bookId);
+    // but there is not migration to do as the bookmark already point to it.
+    ASSERT_EQ(lib->migrateBookmarks(bookId, kiwix::UPGRADE_ONLY), 0);
+
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::ALLOW_DOWNGRADE), bookId);
+}
+
+TEST_F(LibraryTest, GetBestTargetBookIdInvalidOlder)
+{
+    auto bookId = std::string("0c45160e-f917-760a-9159-dfe3c53cdcdd");
+
+    auto book = lib->getBookById(bookId);
+
+    auto invalidBookmark = createBookmark(book);
+    invalidBookmark.setBookId("invalid-book-id");
+    lib->addBookmark(invalidBookmark);
+
+    ASSERT_EQ(lib->getBestTargetBookId(invalidBookmark, kiwix::UPGRADE_ONLY), bookId+"_date");
+    ASSERT_EQ(lib->getBestTargetBookId(invalidBookmark, kiwix::ALLOW_DOWNGRADE), bookId+"_date");
+}
+
+TEST_F(LibraryTest, GetBestTargetBookIdInvalidNewer)
+{
+    auto bookId = std::string("0c45160e-f917-760a-9159-dfe3c53cdcdd");
+
+    auto book = lib->getBookById(bookId);
+    EXPECT_EQ(book.getDate(), "2018-10-08");
+
+    auto invalidBookmark = createBookmark(book);
+    invalidBookmark.setBookId("invalid-book-id");
+    invalidBookmark.setDate("2020-10-08");
+    lib->addBookmark(invalidBookmark);
+
+    ASSERT_EQ(lib->getBestTargetBookId(invalidBookmark, kiwix::UPGRADE_ONLY), "");
+    ASSERT_EQ(lib->getBestTargetBookId(invalidBookmark, kiwix::ALLOW_DOWNGRADE), bookId+"_date");
 }
```

Comment:
When reading the tests explicitly calling the `getBestTargetBookId()` function, `bookId+"_date"` is not intuitive. `bookId+"_updated_1year_later"` or `"newer_version_of_"+bookId` is more self-explaining.

For the current set of tests this comment applies to the `*_flavour` and `*_date_flavour` variants to a much lesser extent, since they are used only in the `LibraryTest.MigrateBookmark` test case. Shouldn't we have tests for `getBestTargetBookId()` involving those combinations too?

Reply:
I have renamed `_date` to `_udpated1yearlater`.

> For the current set of tests this comment applies to the *_flavour and *_date_flavour variants to a much lesser extent, since they are used only in the LibraryTest.MigrateBookmark test case. Shouldn't we have tests for getBestTargetBookId() involving those combinations too?

Indeed. We could. But `MigrateBookmark` is using `getBestTargetBookId` too, so it is indirectly tested.

Reply:
Small test added.
Not a full testing as the method is indirectly tested with `MigrateBookmark` and `getBestTargetBookId(bookName, ...)`.